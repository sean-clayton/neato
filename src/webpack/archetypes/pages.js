import HtmlWebpackPlugin from 'html-webpack-plugin'
import { join } from 'path'
import { optimize } from 'webpack'
import actions from '../../actions'

export default {
  name: 'pages',
  configure({ pages = [], action, projectPath, optimize, vendor = [], filename = {} }) {
    if (pages.length === 0) { return {} }

    const entry = configureEntry(pages, vendor)
    const plugins = configurePlugins(pages, action)

    return {
      output: {
        publicPath: '/',
        path: join(projectPath, 'dist'),
        filename: optimize ? filename.dev || '[name]-[chunkhash].js' : filename.prod || '[name]-[hash].js',
        chunkFilename: optimize ? '[name]-[chunkhash].chunk.js' : '[name]-[hash].chunk.js'
      },
      plugins,
      entry
    }
  }
}

function configureEntry(pages, vendor) {
  let entry = {}

  pages.forEach((page) => {
    if (page === 'vendor') throw new Error('Reserved name "vendor"')
    entry[page] = [`./${page}`]
  })

  entry.vendor = vendor

  return entry
}

function configurePlugins(pages, action) {
  const plugins = pages.map((page) => {
    return new HtmlWebpackPlugin({
      template: `${page}.html`,
      filename: `${page}.html`,
      chunks: ['vendor', page]
    })
  })

  if (action !== actions.TEST) {
    plugins.push(new optimize.CommonsChunkPlugin({ name: 'vendor' }))
  }

  return plugins
}
