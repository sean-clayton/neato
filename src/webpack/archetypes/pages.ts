import HtmlWebpackPlugin from 'html-webpack-plugin'
import { join } from 'path'
import { optimize } from 'webpack'
import actions from '../../actions'

export default {
  name: 'pages',
  configure ({ pages = [], action, projectPath }) {
    if (pages.length === 0) { return {} }

    const entry = configureEntry(pages)
    const plugins = configurePlugins(pages, action)

    return {
      output: {
        path: join(projectPath, 'dist'),
        filename: '[name]-[hash].js',
        chunkFilename: '[name]-[hash].chunk.js'
      },
      plugins,
      entry
    }
  }
}

function configureEntry (pages) {
  let entry = {}

  pages.forEach((page) => {
    entry[page] = [`./${page}`]
  })

  return entry
}

function configurePlugins (pages, action) {
  const plugins = pages.map((page) => {
    return new HtmlWebpackPlugin({
      template: `${page}.html`,
      filename: `${page}.html`,
      chunks: ['common', page]
    })
  })

  if (pages.length > 1 && action !== actions.TEST) {
    plugins.push(new optimize.CommonsChunkPlugin({ name: 'common' }))
  }

  return plugins
}
