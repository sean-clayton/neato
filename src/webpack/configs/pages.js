// @flow

import HtmlWebpackPlugin from 'html-webpack-plugin'
import { optimize } from 'webpack'
import path from 'path'

export default (neatoConfig: NeatoConfigType): Object => {
  const chunks = ['shared', 'vendor', 'webpack']

  const configurePlugins = (): any[] => {
    return [
      ...(neatoConfig.options.pages
        ? neatoConfig.options.pages.map((page: string) => {
          if (chunks.indexOf(page) !== -1) { return new Error(`Reserved name ${page}!`) }
          return new HtmlWebpackPlugin({
            template: `${page}.html`,
            hash: true,
            filename: `${page}.html`
          })
        })
        : []
      ),
      new optimize.CommonsChunkPlugin({
        name: chunks
      })
    ]
  }

  const entry = {
    ...(neatoConfig.options.pages
      ? neatoConfig.options.pages
        .map((page: string) => ({
          [page]: `./${page}`
        }))
        .reduce((a, b) => ({
          ...a,
          ...b
        }))
      : {}),
    vendor: neatoConfig.options.vendor || []
  }

  return neatoConfig.options.pages && neatoConfig.options.pages.length === 0
    ? new Error('Could not find page. Did you forget to set the `pages` option in your neato.config.js file?')
    : {
      output: {
        publicPath: '/',
        path: path.join(neatoConfig.projectPath, 'dist'),
        filename: neatoConfig.optimize ? '[name]-[chunkhash].js' : '[name]-[hash].js',
        chunkFilename: neatoConfig.optimize ? '[name]-[chunkhash].chunk.js' : '[name]-[hash].chunk.js'
      },
      plugins: configurePlugins(),
      entry
    }
}
