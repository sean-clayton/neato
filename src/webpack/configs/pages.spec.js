/* eslint-env jest */

import pages from './pages'

describe('webpack/pages', () => {
  it('works', () => {
    const neatoMock = {
      neatoPath: '/test/neato/path',
      projectPath: '/test/project/path',
      optimize: false,
      options: {
        pages: ['index', 'other-page']
      }
    }
    expect(pages(neatoMock)).toMatchSnapshot()
  })
})
