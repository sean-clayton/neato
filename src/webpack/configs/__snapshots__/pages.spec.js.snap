exports[`webpack/pages works 1`] = `
Object {
  "entry": Object {
    "index": "./index",
    "other-page": "./other-page",
    "vendor": Array [],
  },
  "output": Object {
    "chunkFilename": "[name]-[hash].chunk.js",
    "filename": "[name]-[hash].js",
    "path": "/test/project/path/dist",
    "publicPath": "/",
  },
  "plugins": Array [
    HtmlWebpackPlugin {
      "options": Object {
        "cache": true,
        "chunks": "all",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "index.html",
        "hash": true,
        "inject": true,
        "minify": false,
        "showErrors": true,
        "template": "index.html",
        "title": "Webpack App",
        "xhtml": false,
      },
    },
    HtmlWebpackPlugin {
      "options": Object {
        "cache": true,
        "chunks": "all",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "other-page.html",
        "hash": true,
        "inject": true,
        "minify": false,
        "showErrors": true,
        "template": "other-page.html",
        "title": "Webpack App",
        "xhtml": false,
      },
    },
    CommonsChunkPlugin {
      "async": undefined,
      "chunkNames": Array [
        "shared",
        "vendor",
        "webpack",
      ],
      "filenameTemplate": undefined,
      "ident": "/home/sean/code/neato/node_modules/webpack/lib/optimize/CommonsChunkPlugin.js0",
      "minChunks": undefined,
      "minSize": undefined,
      "selectedChunks": undefined,
    },
  ],
}
`;
